drivers/gpio/gpio_nrfx.c

```diff
@@ -67,10 +67,14 @@ static void gpiote_channel_free(u32_t abs_pin)
 	u32_t intenset = nrf_gpiote_int_is_enabled(NRF_GPIOTE_INT_IN_MASK);
 
 	for (size_t i = 0; i < GPIOTE_CH_NUM; i++) {
+		nrf_gpiote_events_t evt =
+			offsetof(NRF_GPIOTE_Type, EVENTS_IN[i]);
+
 		if ((nrf_gpiote_event_pin_get(i) == abs_pin)
 		    && (intenset & BIT(i))) {
 			nrf_gpiote_event_disable(i);
 			nrf_gpiote_int_disable(BIT(i));
+			nrf_gpiote_event_clear(evt);
```

Comment:
@Mierunski @anangl can you take another look at this?

Reply:
both are off today, @pizi-nordic had a look

Reply:
I am not 100% sure here.  AFAIK, the interrupt  (or some other poll-based-) handler, which processes the corresponding event, is supposed to clear the EVENTS register.

Alternatively, the EVENTS register is cleared when the event is enabled (to make sure we do not have trash value there). 

So this requires some justification, IMO

Reply:
>> I am not 100% sure here. AFAIK, the interrupt (or some other poll-based-) handler, which processes the corresponding event, is supposed to clear the EVENTS register.

The event is cleared in isr handler. The problem is that when application disables interrupt, no disable interrupt should be received after call is finished. It is easy to imagine the problem if we think about irq_disable or spinlock. Application would not be able to function correctly if late interrupts hit it after call disabling the interrupt is done.

Reply:
>> Alternatively, the EVENTS register is cleared when the event is enabled (to make sure we do not have trash value there).

I agree this is a problem but I don't want to focus on it as it does not impact our application. I see potential issue when two submodules would use GPIO pins and actively enable/disable callbacks. The way this driver is implemented will cause interrupts to be constantly interfering with each other.

There is also a bigger bug than that. I don't recall any lock taken before on gpiote_pin_int_cfg is called. There is a chance that when two threads would enable callback on two different pins, scheduling race would assign them to the same channel.

Reply:
If the "late" interrupt hits before line 77, I suppose you have the exact same problem you are trying to solve with line 77, right?

Reply:
@pdunaj it would be nice if you could elaborate in detail, which problem exactly line 77 is solving. It is not clear from the discussion, here. A step-by-step explanation would be appreciated.

So, I am trying to make few points here, fo which I would appreciate your feedback. 
1. This function de-registers the GPIOTE channel and clears the  hardware interrupt source, associated with this channel. After the interrupt is cleared, i.e. the call in line 76 has returned, it should be the responsibility of the driver and/or the application _not_ to inspect the EVENTS register of a channel (in "thread" mode) we have just de-registered. 

2. If the driver inspects the EVENTS register in IRQ Handler mode, then, as you say, there may be race conditions, so line 77 does not solve anything, IMHO.

3. Typically, in other Nordic drivers in Zephyr, the EVENTS registers are cleared right before an peripheral event is activated, and I would very much like all drivers to comply to this rule. If you think that this is bad/incomplete design, I would of course like to listen :)

Reply:
@ioannisg , I am not sure if I can say it clearer. The goal is to not receive the interrupt A after interrupt A was disabled.
In the current state this driver does not provide this functionality.
After patch is applied interrupts are disabled and after call to disable is done no callback of this type is received.

Note that I am not the author of this nor I intend to fix other issues I see in this code. I have other stuff to do but until this code is broken I cannot progress. If you think this can or should be done in a better way please give exact code. I would much appreciate it.

Reply:
@pdunaj: Could you please move `rf_gpiote_event_clear(evt)` before `nrf_gpiote_event_enable(channel)` in gpiote_channel_alloc() and check if this solves your issue? If this will not work, we will dig a bit deeper into this driver.

Selected as best solution during discussion with @ioannisg.

Reply:
I will check it, thanks

Reply:
I have talked with @pizi-nordic off-line. For anybody that follows this I will explain what is the concern being discussed.

All agree that no interrupt should be received after callback is disabled. @ioannisg expects that even if event is not cleared interrupt should not be triggered after it was disabled by `nrf_gpiote_int_disable`. From this perspective interrupt should be tiggered after being re-enabled by `nrf_gpiote_int_enable` (and just before that event should be cleared as this is the way it is always done).

Reply:
Unfortunately this is not what I observe in the driver behavior. Interrupt is hit after I disable callbacks and before I re-enable them.
I did the change proposed above by @pizi-nordic and it changed nothing. Below is the log after the change.

`bp x` is displayed in callback (interrupt received from driver). `1` on entry, `2` after callbacks are disabled, `3` at exit.
`disable n` is added close to `int_disable` in `gpiote_channel_free` - n is channel number
`enable n` is added close to `int_enable` in `gpiote_channel_alloc` - n is channel number

As is visible in logs interrupts are disabled but I am being hit by one.

```
e: module_state_event module:main state:ready
enable 0
enable 1
enable 2
enable 3
e: module_state_event module:board state:ready
e: module_state_event module:wheel state:ready
e: module_state_event module:buttons state:ready
[bt] [INF] hci_vs_init: HW Platform: Nordic Semiconductor (0x0002)
[bt] [INF] hci_vs_init: HW Variant: nRF52x (0x0002)
[bt] [INF] hci_vs_init: Firmware: Standard Bluetooth controller (0x00) Version 1.12 Build 99
[bt] [WRN] bt_init: No ID address. Expecting one to come from storage.
[bt] [INF] bt_dev_show_info: Identity: ca:74:81:0b:25:1f (random)
[bt] [INF] bt_dev_show_info: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0x05f1
[bt] [INF] bt_dev_show_info: LMP: version 5.0 (0x09) subver 0xffff
e: module_state_event module:ble_state state:ready
e: module_state_event module:bas state:ready
e: module_state_event module:dis state:ready
e: module_state_event module:hids state:ready
e: module_state_event module:ble_adv state:ready




bp 1
dissable 0
dissable 1
dissable 2
dissable 3
bp 2
bp 3
bp 1
bp 2
ASSERTION FAIL [!atomic_get(&scanning)] @ /home/padu/work/fw-nrfconnect-nrf/samples/nrf_desktop/src/hw_interface/buttons_pca20041.c:184:
```